ws/types.nim

```diff
@@ -96,13 +95,21 @@ type
 
   Ext* = ref object of RootObj
     name*: string
-    options*: Table[string, string]
     session*: WSSession
 
-  ExtFactory* = proc(
-    name: string,
-    session: WSSession,
-    options: Table[string, string]): Ext {.raises: [Defect].}
+  ExtParam* = object
+    name* : string
+    value*: string
+
+  ExtFactoryProc* = proc(
+    isServer: bool,
+    args: seq[ExtParam]): Result[Ext, string] {.
+      gcsafe, raises: [Defect].}
+
+  ExtFactory* = object
+    name*: string
+    factory*: ExtFactoryProc
+    clientOffer*: string
```

Comment:
Just curious, why do we need this?

Reply:
1. `openArray[(string, ExtFactoryProc)]` does not work for closure proc, we need closure proc.
2. why we need closure proc? we need to capture extension configuration from user
3. keep it ergonomics
4. `clientOffer` is needed before the actual extension creation. the server also can rejects or accepts with configuration different from client offered.

```nim
proc extFactoryMaker(offerPadding: bool, useStdLib: bool): ExtFactory =
   proc factoryProc(.....): Ext =
        # capture extension configuration
        # e.g. `useStdLib`
        
   ExtFactory (
      name : "base64",
      factory: factoryProc,
      clientOffer: "base64; pad=" & $offerPadding  # client configuration
   )
```

Reply:
I was mostly asking about `clientOffer`, but thanks for the broader explanation.